### PR TITLE
fix: variant price update

### PR DESCRIPTION
**Pull request by zakariaelas**
### What

- fixes updating a variant's prices

### How

- The `setCurrencyPrice` method used the `IsNull` typeorm operator which, for some reason, did not work as expected. The solution was to use query builder and directly query for `money_amount.region_id IS NULL`. 
- The logic for updating variant prices did not account for deleting old prices. It only added/updated prices. The solution was to fetching old or "obsoletePrices" and removing them from the MoneyAmount table. This is implemented by fetching all money amounts belonging to a variant id AND NOT matching any price in the `prices` payload. 

### Testing

- integration test for changing an existing price
- integration test for changing an existing price (this time given a `region_id`)
- integration test for adding a new price
- integration test for replacing a price (or in other words deleting old price, then adding a new one)
- integration test for replacing + adding a price
- integration test for updating (this time given a `region_id`) + adding a price
- unit test for `ProductVariantService.updateVariantPrices` to test removal of obsolete prices + addition/update of new ones using a currency_code
- unit test for `ProductVariantService.updateVariantPrices` to test removal of obsolete prices + addition/update of new ones using a region_id
- adjust unit test for `ProductVariantService.update`

### CODE REVIEW

File: packages/medusa/src/services/product-variant.ts [hunk 1]
```diff
@@ -333,6 +317,48 @@ class ProductVariantService extends BaseService {
...
+          new Brackets((qb) => {
+            qb.where("money_amount.currency_code NOT IN (:...currency_codes)", {
+              currency_codes: prices.map((p) => p.currency_code),
+            }).orWhere("money_amount.region_id NOT IN (:...region_ids)", {
+              region_ids: prices.map((p) => p.region_id),
+            })
+          })
+        )
+        .getMany()
+
```
Comment (srindom) [inline packages/medusa/src/services/product-variant.ts:345]: This is brilliant - one thing I would like is that this gets moved into the `MoneyAmountRepository` as a custom repo method. Generally, we should avoid having query builder stuff in the service layer.

Also I think some of the queries here can be TypeORM-ified more: e.g. you can do:
```ts
qb.where({
  currency_code: Not(In(prices.map(p => p.currency_code)))
})
```

Comment (srindom) [reply packages/medusa/src/services/product-variant.ts:345]: Also `.where({ variant_id: variantId })`

File: packages/medusa/src/services/product-variant.ts [hunk 2]
```diff
@@ -348,13 +374,16 @@ class ProductVariantService extends BaseService {
...
+          currency_code: price.currency_code,
+        })
+        .andWhere("money_amount.variant_id = :variant_id", {
           variant_id: variantId,
-          region_id: IsNull(),
-        },
-      })
+        })
+        .andWhere("money_amount.region_id IS NULL")
+        .getOne()
```
Comment (srindom) [inline packages/medusa/src/services/product-variant.ts:386]: Let's also move this into the repo. We can potentially make it even more powerful so it functions more like an upsert in the repository layer and then we can name the function something like: `upsertCurrencyPrice`? This way we don't have to do the query and decide pattern in the service layer which gets a bit tedious.

**Review by srindom — CHANGES_REQUESTED** [review]

File: packages/medusa/src/repositories/money-amount.ts [hunk 3]
```diff
@@ -1,5 +1,55 @@
...
 @EntityRepository(MoneyAmount)
-export class MoneyAmountRepository extends Repository<MoneyAmount> { }
+export class MoneyAmountRepository extends Repository<MoneyAmount> {
+  public async findVariantPricesNotIn(
+    variantId: string,
+    prices: Price[]
+  ) {
+    const pricesNotInPricesPayload = await this.createQueryBuilder(
+    )
+      .where({
```
Comment (srindom) [inline packages/medusa/src/repositories/money-amount.ts:14]: nit: Don't know if your linter is acting up here?? :D

**Review by srindom — APPROVED** [review]
LGTM w. comments; seems to be a small linter issue please fix and merge :)

File: packages/medusa/src/repositories/money-amount.ts (hunk 3 above)
Comment (zakariaelas) [reply packages/medusa/src/repositories/money-amount.ts]: it was :sob:

**Pull request merged**